apply/v2/scale.go

```diff
@@ -0,0 +1,165 @@
+// Copyright © 2021 Alibaba Group Holding Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package apply
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/alibaba/sealer/apply/v2/applytype"
+	"github.com/alibaba/sealer/common"
+	"github.com/alibaba/sealer/logger"
+	v2 "github.com/alibaba/sealer/types/api/v2"
+	"github.com/alibaba/sealer/utils"
+)
+
+// NewScaleApplierFromArgs will filter ip list from command parameters.
+func NewScaleApplierFromArgs(clusterfile string, scaleArgs *common.RunArgs, flag string) (applytype.Interface, error) {
+	cluster := &v2.Cluster{}
+	if err := utils.UnmarshalYamlFile(clusterfile, cluster); err != nil {
+		return nil, err
+	}
+	if scaleArgs.Nodes == "" && scaleArgs.Masters == "" {
+		return nil, fmt.Errorf("the node or master parameter was not committed")
+	}
+
+	var err error
+	switch flag {
+	case common.JoinSubCmd:
+		err = Join(cluster, scaleArgs)
+	case common.DeleteSubCmd:
+		err = Delete(cluster, scaleArgs)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	if err := utils.MarshalYamlToFile(clusterfile, cluster); err != nil {
+		return nil, err
+	}
+	applier, err := NewApplier(cluster)
+	if err != nil {
+		return nil, err
+	}
+	return applier, nil
+}
+
+func Join(cluster *v2.Cluster, scalingArgs *common.RunArgs) error {
+	/*	switch cluster.Spec.Provider {
+		case common.BAREMETAL:
+			return joinBaremetalNodes(cluster, scalingArgs)
+		case common.AliCloud:
+			return joinInfraNodes(cluster, scalingArgs)
+		case common.CONTAINER:
+			return joinInfraNodes(cluster, scalingArgs)
+		default:
+			return fmt.Errorf(" clusterfile provider type is not found ！")
+		}*/
+	return joinBaremetalNodes(cluster, scalingArgs)
+}
+
+func joinBaremetalNodes(cluster *v2.Cluster, scaleArgs *common.RunArgs) error {
+	if err := PreProcessIPList(scaleArgs); err != nil {
+		return err
+	}
+	if (!IsIPList(scaleArgs.Nodes) && scaleArgs.Nodes != "") || (!IsIPList(scaleArgs.Masters) && scaleArgs.Masters != "") {
+		return fmt.Errorf(" Parameter error: The current mode should submit iplist！")
+	}
+	if scaleArgs.Masters != "" && IsIPList(scaleArgs.Masters) {
+		for i := 0; i < len(cluster.Spec.Hosts); i++ {
+			role := cluster.Spec.Hosts[i].Roles
+			//if a separate `node` role exists for host, add ip to host
```

Comment:
if a separate `node` role exists for host, add ip to host;
if not, create a new host